pytorch_lightning/callbacks/gpu_stats_monitor.py

```diff
@@ -0,0 +1,164 @@
+        if self._log_stats.gpu_utilization:
+            self._log_gpu(trainer)
+        if self._log_stats.memory_utilization:
+            self._log_memory(trainer)
+
+        if self._log_stats.inter_step_time and self.snap_inter_step_time:
+            # First log at beginning of second step
+            trainer.logger.log_metrics(
+                {'batch_time/inter_step (ms)': (time.time() - self.snap_inter_step_time) * 1000},
+                step=trainer.global_step
+            )
+
+        if self._log_stats.intra_step_time:
+            self.snap_intra_step_time = time.time()
+
+    def on_train_batch_end(self, trainer, pl_module, batch, batch_idx, dataloader_idx):
+        if self._log_stats.gpu_utilization:
+            self._log_gpu(trainer)
+
+        if self._log_stats.memory_utilization:
+            self._log_memory(trainer)
+
+        if self._log_stats.fan_speed:
+            trainer.logger.log_metrics(self._get_gpu_stat("fan.speed", "%"), step=trainer.global_step)
+
+        if self._log_stats.temperature:
+            trainer.logger.log_metrics(self._get_gpu_stat("temperature.gpu", "degrees C"), step=trainer.global_step)
+            trainer.logger.log_metrics(self._get_gpu_stat("temperature.memory", "degrees C"), step=trainer.global_step)
+
+        if self._log_stats.inter_step_time:
+            self.snap_inter_step_time = time.time()
+
+        if self._log_stats.intra_step_time and self.snap_intra_step_time:
+            trainer.logger.log_metrics(
+                {'batch_time/intra_step (ms)': (time.time() - self.snap_intra_step_time) * 1000},
+                step=trainer.global_step
+            )
+
+    def on_train_start(self, trainer, pl_module):
+        if not trainer.logger:
+            raise MisconfigurationException(
+                'Cannot use GPUStatsMonitor callback with Trainer that has no logger.'
+            )
+
+        if not trainer.on_gpu:
+            rank_zero_warn(
+                'You are using GPUStatsMonitor but are not running on GPU.'
+                ' Logged utilization will be independent from your model.', RuntimeWarning
+            )
+
+    def on_train_epoch_start(self, trainer, pl_module):
+        self.snap_intra_step_time = None
+        self.snap_inter_step_time = None
+
+    @staticmethod
+    def _get_gpu_stat(pitem: str, unit: str):
+        result = subprocess.run(
+            [shutil.which("nvidia-smi"), f"--query-gpu={pitem}", "--format=csv,nounits,noheader"],
+            encoding="utf-8",
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,  # for backward compatibility with python version 3.6
+            check=True
+        )
+
+        try:
+            gpu_usage = [float(x) for x in result.stdout.strip().split(os.linesep)]
+        except ValueError:
+            gpu_usage = [0]
+
+        return {f"gpu_{pitem}/gpu_id_{index} ({unit})": usage for index, usage in enumerate(gpu_usage)}
+
+    def _log_gpu(self, trainer):
```

Comment:
```suggestion
    def _log_usage(self, trainer):
```
as it is all about GPU, `_log_gpu` does not say much...